### DISCUSSION TITLE
Welcome to rust-music-theory Discussions!

### DISCUSSION THREAD

**Discussion by ozankasikci** [ozankasikci/fantasy-in-rust#21] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by henryksloan** [comment]
Hi! I'm Henry, and I've been contributing to this project for about a month. I've been having a good time making tweaks and adding useful features. I'm excited and interested in a potential frontend for this library. With some ingenuity and brainstorming. I think we could realistically create a pretty ambitious music engraving/education tool.

To that end, I'll just give some of my recent thoughts about potential approaches to a long-term application. @ozankasikci has mentioned a pure-Rust stack, and I think that's totally viable. However, I think the best way forward for a frontend is to brainstorm options for features, platform, and purpose.

I think if a pure-Rust stack is used, then wrangling with the Rust ecosystem would be the primary technical challenge of the project. That said, it would create interest among a very talented community, and would open the door to technical development and entirely novel features. I could see a sheet music rendering library being developed using 1) a canvas component in `conrod` or something, or 2) simple shape painting in a game library like `Amethyst`, `ggez`, or `Bevy`. 

My recommended alternative would be a web stack, including Electron and possibly WASM. [This blog post](https://keminglabs.com/blog/building-a-fast-electron-app-with-rust/) uses Electron and [neon](https://github.com/neon-bindings/neon) to implement a frontend using Electron as a thin frontend layer. This would allow us to leverage [Vexflow](https://www.vexflow.com/) and [OpenSheetMusicDisplay](https://opensheetmusicdisplay.org/), as well as better Midi and MusicXML support. Honestly, I think a webapp is the right approach if the goal is education. There is also more room for improvement in the space of web-based composition tools than in that of desktop engraving software.

On my end, I'm developing on a library called Excelsis (I'll make the repo public when it's ready), which is currently a collection of features that wrap this library. I have a Rules Engine for ensuring adherence to counterpoint rules between any number of voices, and a generator which can produce counterpoint harmonies above or below a given melody ([see some early examples here](https://musescore.com/user/27207526/scores/6299932/s/NhFtiy)). I have also implemented a simple playback API for rust-music-theory using `rodio`. If a future application could use it, I could refactor some of these features to work with the application.

**Reply by ozankasikci** [reply]
Hi Henry, thanks for the introduction!

I think we definitely need an awesome frontend for this awesome project, however, we first need to define the scope, purpose, and style of the frontend before writing any code for that. We can start a new discussion for the frontend project and start discussing these points there.

Whether the frontend would be implemented in Rust or in some other technology is inherently depending on the purpose and the features that we're going to choose to add. I've actually read the blog post you mentioned but never tried it in action, could be an option. But WASM is also definitely worth a try, we could create a web-based but performant application that can be used by anyone without installing/downloading anything, like this idea!

I believe you have already been a valuable contributor to the project and we'll bring this library to a great point with collective effort and ambition!
